File: LeetCode/Python Solutions/Maximum Profit in Job Scheduling/maximum.py
from typing import List
from bisect import bisect_left

class Solution:
    def jobScheduling(self, startTime: List[int], endTime: List[int], profit: List[int]) -> int:
        jobs = sorted(zip(startTime, endTime, profit), key=lambda x: x[1])
        dp = [(0, 0)]  # (endTime, profit) pair, initial dummy job with 0 profit

        for start, end, profit in jobs:
            # Find the last job that ends before the current job starts
            i = bisect_left(dp, (start, float('inf'))) - 1
            if dp[i][1] + profit > dp[-1][1]:
                # If this job plus the best previous job is better than the last best job
                dp.append((end, dp[i][1] + profit))

        return dp[-1][1]

File: LeetCode/Java Solutions/Maximum Profit in Job Scheduling/maximum.java
import java.util.Arrays;
import java.util.TreeMap;

class Solution {
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        int n = startTime.length;
        int[][] jobs = new int[n][3];
        for (int i = 0; i < n; i++) {
            jobs[i] = new int[]{startTime[i], endTime[i], profit[i]};
        }
        Arrays.sort(jobs, (a, b) -> a[1] - b[1]);
        
        // TreeMap to store the max profit up to the current time
        TreeMap<Integer, Integer> dp = new TreeMap<>();
        dp.put(0, 0); // Initial value

        for (int[] job : jobs) {
            int curProfit = dp.floorEntry(job[0]).getValue() + job[2];
            if (curProfit > dp.lastEntry().getValue()) {
                dp.put(job[1], curProfit);
            }
        }
        
        return dp.lastEntry().getValue();
    }
}
